dell/src/main/java/org/apache/iceberg/dell/ecs/EcsCatalog.java

```diff
@@ -0,0 +1,511 @@
+  public EcsCatalog() {
+  }
+
+  @Override
+  public void initialize(String name, Map<String, String> properties) {
+    this.catalogName = name;
+    this.warehouseLocation = new EcsURI(properties.get(CatalogProperties.WAREHOUSE_LOCATION));
+    this.client = DellClientFactories.from(properties).ecsS3();
+    this.fileIO = initializeFileIO(properties);
+
+    this.closeableGroup = new CloseableGroup();
+    closeableGroup.addCloseable(client::destroy);
+    closeableGroup.addCloseable(fileIO);
+    closeableGroup.setSuppressCloseFailure(true);
+  }
+
+  private FileIO initializeFileIO(Map<String, String> properties) {
+    String fileIOImpl = properties.get(CatalogProperties.FILE_IO_IMPL);
+    if (fileIOImpl == null) {
+      FileIO io = new EcsFileIO();
+      io.initialize(properties);
+      return io;
+    } else {
+      return CatalogUtil.loadFileIO(fileIOImpl, properties, hadoopConf);
+    }
+  }
+
+  @Override
+  protected TableOperations newTableOps(TableIdentifier tableIdentifier) {
+    return new EcsTableOperations(String.format("%s.%s", catalogName, tableIdentifier),
+        tableURI(tableIdentifier), fileIO, this);
+  }
+
+  @Override
+  protected String defaultWarehouseLocation(TableIdentifier tableIdentifier) {
+    String tableName = tableIdentifier.name();
+    StringBuilder sb = new StringBuilder();
+
+    sb.append(warehouseLocation).append('/');
+    for (String level : tableIdentifier.namespace().levels()) {
+      sb.append(level).append('/');
+    }
+    sb.append(tableName);
+
+    return sb.toString();
+  }
+
+  /**
+   * Iterate all table objects with the namespace prefix.
+   */
+  @Override
+  public List<TableIdentifier> listTables(Namespace namespace) {
+    if (!namespace.isEmpty() && !namespaceExists(namespace)) {
+      throw new NoSuchNamespaceException("Namespace %s does not exist", namespace);
+    }
+
+    String marker = null;
+    List<TableIdentifier> results = Lists.newArrayList();
+    EcsURI prefix = namespacePrefix(namespace);
+    do {
+      ListObjectsResult listObjectsResult = client.listObjects(
+          new ListObjectsRequest(prefix.bucket())
+              .withDelimiter("/")
+              .withPrefix(prefix.name())
+              .withMarker(marker));
+      marker = listObjectsResult.getNextMarker();
+      results.addAll(listObjectsResult.getObjects().stream()
+          .filter(s3Object -> s3Object.getKey().endsWith(TABLE_OBJECT_SUFFIX))
+          .map(object -> parseTableId(namespace, prefix, object))
+          .collect(Collectors.toList()));
+    } while (marker != null);
+
+    LOG.debug("Listing of namespace: {} resulted in the following tables: {}", namespace, results);
+    return results;
+  }
+
+  /**
+   * Get object prefix of namespace.
+   */
+  private EcsURI namespacePrefix(Namespace namespace) {
+    String prefix;
+    if (namespace.isEmpty()) {
+      prefix = warehouseLocation.name();
+    } else {
+      prefix = String.format("%s%s/", warehouseLocation.name(),
```

Comment:
Will the warehouseLocation has a slash `\`  as the last char ?  If so, then we may need to trim the last slash before constructing the string formatter.